### PR TITLE
Mp4 to h264 and acc seperated class

### PR DESCRIPTION
**Pull request by mortenvp**

### CODE REVIEW

File: src/petro/extractor/mpeg_versions.hpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Steinwurf ApS 2016.
+// All Rights Reserved
+//
+// Distributed under the "BSD License". See the accompanying LICENSE.rst file.
+namespace petro
+{
+    enum mpeg_versions
```
Comment (mortenvp) [inline src/petro/extractor/mpeg_versions.hpp]: Use an `enum class` for strongly typed enums.

**Comment by petya2164** [comment]
@looopTools I will fix the style issues here and I will add some functionality to make these classes actually useful in wurf-it. We need to query the decoding and presentation timestamps for the current sample and we need looping.

**Comment by petya2164** [comment]
I made some significant simplifications here. The objective is creating an extractor API that covers all the required functionality in wurf-it, wurf-gstreamer and petro-android. I will check if it works everywhere before we merge this.
So take a look, and perhaps we can find a better name for `advance_to_next_sample()`.

**Comment by looopTools** [comment]
I do not understand why this: https://github.com/steinwurf/petro/blob/mp4_to_h264_and_acc_seperated_class/src/petro/extractor/aac_extractor.cpp#L110 and the method sample_data() starting at line 165 needs to be to seperated functions. I was asked to implement the functionality, so it was similar to a iterator there for the `has_next` and `next` method.  I do not understand the advantage over the current approach, contra the one I implemented, can you elaborate on this @petya2164 ?

**Comment by jpihl** [comment]
I like the advance stuff. It moves all the complex stuff into a single fuction and simplifies the getter for the sample.
But let's call the function "advance".

**Comment by mortenvp** [comment]
I would suggest an approach where the API for iterating though samples would be, something like:

```
/// @return true if there are no more samples
bool at_end() const;

/// Go to the next sample
void advance();

/// @return Sample data of the current sample
std::vector<uint8_t> sample_data();
```

So one function for querying where there is more data (it has no side effects), one function for advancing the extraction point, and one for getting the data.

**Comment by petya2164** [comment]
@mortenvp That is an interesting suggestion. I'm not sure if the separation of at_end and advance is actually helpful in this case. Do you have a concrete scenario where you would not call advance right after at_end?

**Comment by petya2164** [comment]
I should mention that the Java-style has_next+next iterator is actually an "alien idiom" in C++, this is yet another Java feature that I hate :)
A normal C++ iterator only has advance(++), and then you dereference(*) to access the current data. This is the most basic example that I could find: http://www.cplusplus.com/reference/iterator/iterator/
Here we need to access multiple pieces of data (sample data + timestamps) that belong to the very same sample, so we cannot advance the extraction in the getter methods. Otherwise you might get a timestamp that belongs to the next or previous sample.

**Comment by petya2164** [comment]
I added the at_end() method which helps to reduce the EOS-related code in wurf-it and wurf-gstreamer.
The iteration is still done with advance_to_next_sample, I think this is the simplest solution for our use case.

**Comment by petya2164** [comment]
I added some getters for the audio properties (these are needed in petro-android). With these last changes, I confirmed that the API covers all requirements in wurf-it, wurf-gstreamer and petro-android (and we can remove a lot of duplicated code). If there is a new requirement in the future, then we should implement it here to avoid this blatant code duplication, actually triplication in this case :). 

Any final objections against merging this?

**Comment by jpihl** [comment]
I would prefer to just call the fuction ``advance``, since it's not only the sample data which is being advanced. It's all or most of the sample's attributes and data.

**Comment by mortenvp** [comment]
@petya2164 The main motivation is readability. and ensuring that functions have only one purpose. 
So I think it becomes more flexible that way, but also more easy to understand what is happening.

The C++ iterators have three operations: 

1) Advance `operator++` like the proposed `advance` function here.
2) Comparison `operator !=` like the proposed `at_end` function here.
3) Extraction `operator*` like the proposed `sample_data` and friends.

I would probably not have both and `at_end` and an `advance` that returns `bool` it should be one or the other IMO. 

This can also be used in e.g. a `for` loop if that is prettier:

```
for(; !extractor.at_end(); extractor.advance())
```

**Comment by looopTools** [comment]
So you want `advance` to have return type `void`? Such that we almost ends up with the has next approach?

**Comment by mortenvp** [comment]
Yes, in my opinion `advance` should return `void` and `assert` if called when no more samples are available.

**Comment by looopTools** [comment]
I agree with Morten on this

**Comment by petya2164** [comment]
@mortenvp Yes, with that solution we would get back to the has_next approach :( :(
I should point out that the for-cycle does not work, because `advance` would be called at the *end* of a cycle. As the first sample is not loaded automatically, you would fail to read any data in the first cycle. 
A valid cycle under your assumptions would be like this:
```
while (!extractor.at_end())
{
    extractor.advance();
    // Use the extracted data
}
```
This looks strange in C++, as the user has to remember that `advance` must be the first method in the cycle. I think this is error-prone (especially if you compare it to the combined approach).
Bottom line: readability is worse in this case, as we have separated 2 functions unnecessarily. I consider `at_end` as a helper function which is purely optional.
We could discuss the various options for ages, but it would be better to move on to the more important features (this is actually blocking me now).

**Comment by petya2164** [comment]
So we are thinking too much about enforcing an iterator pattern over this extraction, and the name `advance` certainly gives the wrong impression here.
If we call the function `load_next_sample`, then we get closer to its actual purpose. Then it is totally natural to return a bool to indicate if the load operation was successful or not.
Do you guys agree that it's a more accurate name?

**Comment by petya2164** [comment]
I will merge this now since there were no objections to my last proposal.
We can always add new functionality if something is missing.
